pkg/controller/kubemon/statefulset.go

```diff
@@ -0,0 +1,385 @@
+	}
+}
+
+func buildServiceAccountName(instance *v1alpha1.DynaKube) string {
+	if instance.Spec.KubernetesMonitoringSpec.ServiceAccountName != "" {
+		return instance.Spec.KubernetesMonitoringSpec.ServiceAccountName
+	}
+	return MonitoringServiceAccount
+}
+
+func buildAffinity() *corev1.Affinity {
+	return &corev1.Affinity{
+		NodeAffinity: buildNodeAffinity(),
+	}
+}
+
+func buildNodeAffinity() *corev1.NodeAffinity {
+	return &corev1.NodeAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution: buildNodeSelectorForAffinity(),
+	}
+}
+
+func buildNodeSelectorForAffinity() *corev1.NodeSelector {
+	return &corev1.NodeSelector{
+		NodeSelectorTerms: []corev1.NodeSelectorTerm{
+			{MatchExpressions: buildKubernetesBetaArchExpression()},
+			{MatchExpressions: buildKubernetesArchExpression()},
+		},
+	}
+}
+
+func buildKubernetesBetaArchExpression() []corev1.NodeSelectorRequirement {
+	return buildKubernetesExpression(KubernetesBetaArch, KubernetesBetaOs)
+}
+
+func buildKubernetesArchExpression() []corev1.NodeSelectorRequirement {
+	return buildKubernetesExpression(KubernetesArch, KubernetesOs)
+}
+
+func buildKubernetesExpression(archKey string, osKey string) []corev1.NodeSelectorRequirement {
+	return []corev1.NodeSelectorRequirement{
+		{
+			Key:      archKey,
+			Operator: corev1.NodeSelectorOpIn,
+			Values:   []string{AMD64, ARM64},
+		},
+		{
+			Key:      osKey,
+			Operator: corev1.NodeSelectorOpIn,
+			Values:   []string{LINUX},
+		},
+	}
+}
+
+func buildVolumes(instance *v1alpha1.DynaKube) []corev1.Volume {
+	return append([]corev1.Volume{}, buildCustomPropertiesVolume(instance)...)
+}
+
+func buildCustomPropertiesVolume(instance *v1alpha1.DynaKube) []corev1.Volume {
+	customProperties := instance.Spec.KubernetesMonitoringSpec.CustomProperties
+	if isCustomPropertiesNilOrEmpty(customProperties) {
+		return nil
+	}
+	return []corev1.Volume{{
+		Name: customproperties.VolumeName,
+		VolumeSource: corev1.VolumeSource{
+			Secret: buildSecretVolumeSource(
+				buildSecretName(instance),
+				customproperties.DataKey,
+				customproperties.DataPath,
+			),
+		},
+	},
+	}
+}
+
+func isCustomPropertiesNilOrEmpty(customProperties *v1alpha1.DynaKubeValueSource) bool {
+	return customProperties == nil ||
+		(customProperties.Value == "" &&
+			customProperties.ValueFrom == "")
+}
+
+func buildSecretVolumeSource(name string, key string, path string) *corev1.SecretVolumeSource {
+	return &corev1.SecretVolumeSource{
+		SecretName: name,
+		Items: []corev1.KeyToPath{
+			{Key: key, Path: path},
+		},
+	}
+}
+
+func buildSecretName(instance *v1alpha1.DynaKube) string {
+	valueFrom := instance.Spec.KubernetesMonitoringSpec.CustomProperties.ValueFrom
+	if valueFrom == "" {
+		valueFrom = fmt.Sprintf("%s-kubernetes-monitoring%s", instance.Name, customproperties.Suffix)
+	}
+	return valueFrom
+}
+
+func buildEnvs(instance *v1alpha1.DynaKube, tenantInfo *dtclient.TenantInfo, kubeSystemUID types.UID) []corev1.EnvVar {
```

Comment:
All the logic should be inside of this function.

Reply:
merged build*Envs functions